dm/worker/worker.go

```diff
@@ -49,62 +51,62 @@ type Worker struct {
 	ctx    context.Context
 	cancel context.CancelFunc
 
-	cfg         *Config
-	subTasks    map[string]*SubTask
+	cfg *Config
+
+	subTasks map[string]*SubTask
+
 	relayHolder *RelayHolder
 	relayPurger *purger.Purger
 
-	meta   *FileMetaDB
+	meta   *Metadata
+	db     *leveldb.DB
 	tracer *tracing.Tracer
 }
 
 // NewWorker creates a new Worker
-func NewWorker(cfg *Config) *Worker {
-	w := Worker{
+func NewWorker(cfg *Config) (*Worker, error) {
+	w := &Worker{
 		cfg:         cfg,
-		subTasks:    make(map[string]*SubTask),
 		relayHolder: NewRelayHolder(cfg),
+		tracer:      tracing.InitTracerHub(cfg.Tracer),
+		subTasks:    make(map[string]*SubTask),
 	}
 
+	// initial relay purger
 	operators := []purger.RelayOperator{
 		w.relayHolder,
 		streamer.GetReaderHub(),
 	}
 	interceptors := []purger.PurgeInterceptor{
-		&w,
+		w,
 	}
 	w.relayPurger = purger.NewPurger(cfg.Purge, cfg.RelayDir, operators, interceptors)
-	w.tracer = tracing.InitTracerHub(cfg.Tracer)
-
-	w.closed.Set(closedTrue) // not start yet
-	w.ctx, w.cancel = context.WithCancel(context.Background())
-	return &w
-}
 
-// Init initializes the worker
-func (w *Worker) Init() error {
+	// initial relay holder
 	err := w.relayHolder.Init()
 	if err != nil {
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 
-	InitConditionHub(w)
-
-	w.meta, err = NewFileMetaDB(w.cfg.MetaDir)
+	// open kv db
+	dbDir := path.Join(w.cfg.MetaDir, "kv")
+	w.db, err = openDB(dbDir, defaultKVConfig)
 	if err != nil {
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 
-	return nil
-}
+	// initial metadata
+	w.meta, err = NewMetadata(dbDir, w.db)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
 
-// Start starts working
-func (w *Worker) Start() {
-	w.closed.Set(closedFalse)
-	w.wg.Add(1)
-	defer w.wg.Done()
+	InitConditionHub(w)
 
-	log.Info("[worker] start running")
+	err = w.restoreSubTask()
+	if err != nil {
+		return nil, errors.Trace(err)
```

Comment:
if `restoreSubTask ` fails here, should we need to release resources acquired before? like close the KV DB.

Reply:
it's a part of startup logic, it seems also ok, letm see